src/asami/entities.cljc

```diff
@@ -174,20 +174,32 @@
                                (if-let [overflow (:overflow (ex-data e))]
                                  (reduced last-result)
                                  (throw e))))
-                           (if (and (seqable? obj)
+                           (if (and (seqable? obj)          ; would it be ok performance-wise to turn this into if-let with destructuring to extract e a v to use below instead of (nth ...)?
                                     (= 4 (count obj))
                                     (= :db/add (first obj)))
                              (or
-                              (when (= (nth obj 2) :db/id)
+                               ;; Ex.: `[:db/add [:id X] :id X]` that creates a new entity
+                               (when-let [ref (and (writer/lookup-ref? (nth obj 1))
+                                                   (= (first (nth obj 1)) (nth obj 2))
+                                                   (nth obj 1))]
+                                 (let [new-id (or #_(ids ref) (node/new-node graph))] ; I cannot see where `(ids ref)` would be non-nil in any well-formed tx
```

Comment:
The first rule of data that you didn't create is that it is never well-formed :)